crates/nu-cli/src/commands/cal.rs

```diff
@@ -67,80 +67,164 @@ pub fn cal(args: CommandArgs, registry: &CommandRegistry) -> Result<OutputStream
-                &tag,
-                year_value as i32,
-                current_year,
-                current_month,
-                day_value,
-                &args,
-            );
+            (1, 12)
         } else {
-            let (day_start_offset, number_of_days_in_month, _) =
-                get_month_information(current_year, current_month, current_year);
-
-            add_month_to_table(
-                &mut calendar_vec_deque,
-                &tag,
-                current_year,
-                current_month,
-                Some(current_day),
-                day_start_offset,
-                number_of_days_in_month as usize,
-                &args,
-            );
-        }
+            (current_month, current_month)
+        };
 
-        for item in calendar_vec_deque {
-            yield ReturnSuccess::value(item);
+        let add_months_of_year_to_table_result = add_months_of_year_to_table(
+            &args,
+            &mut calendar_vec_deque,
+            &tag,
+            selected_year,
+            month_range,
+            current_month,
+            current_day_option,
+        );
+
+        match add_months_of_year_to_table_result {
+            Ok(()) => {
+                for item in calendar_vec_deque {
+                    yield ReturnSuccess::value(item);
+                }
+            }
+            Err(error) => yield Err(error),
         }
     };
 
     Ok(stream.to_output_stream())
 }
 
+fn get_invalid_year_shell_error(year_tag: &Tag) -> ShellError {
+    ShellError::labeled_error("The year is invalid", "invalid year", year_tag)
+}
+
+struct MonthHelper {
+    day_number_month_starts_on: usize,
+    number_of_days_in_month: usize,
+    selected_year: i32,
+    selected_month: u32,
+}
+
+impl MonthHelper {
+    pub fn new(selected_year: i32, selected_month: u32) -> Result<MonthHelper, ()> {
+        let mut month_helper = MonthHelper {
+            day_number_month_starts_on: 0,
+            number_of_days_in_month: 0,
+            selected_year,
+            selected_month,
+        };
+
+        let chosen_date_result_one = month_helper.update_day_number_month_starts_on();
+        let chosen_date_result_two = month_helper.update_number_of_days_in_month();
+
+        if chosen_date_result_one.is_ok() && chosen_date_result_two.is_ok() {
+            return Ok(month_helper);
+        }
+
+        Err(())
+    }
+
+    pub fn get_month_name(&self) -> String {
+        let month_name = match self.selected_month {
+            1 => "january",
+            2 => "february",
+            3 => "march",
+            4 => "april",
+            5 => "may",
+            6 => "june",
+            7 => "july",
+            8 => "august",
+            9 => "september",
+            10 => "october",
+            11 => "november",
+            _ => "december",
+        };
+
+        month_name.to_string()
+    }
+
+    fn update_day_number_month_starts_on(&mut self) -> Result<(), ()> {
```

Comment:
If the date is bad, you should still be able to return an error here.  I'd still return ShellError. If you want, you can use something like:

```
ShellError::untagged_runtime_error(...)
```

This lets you create a ShellError, even when you don't have a tag to pass it.